framework/source/class/qx/io/remote/Rpc.js

```diff
@@ -1072,9 +1102,320 @@ qx.Class.define("qx.io.remote.Rpc",
+                  "Invalid jsonrpc data: Unknown request id",
+                  {
+                    request : data,
+                    response: response
+                  }
+                );
+                return reject(ex);
+              }
+              // remove the entry from the queue
+              this.__requestQueue.splice(this.__requestQueue.indexOf(originalRequest), 1);
+              // we have an error
+              if (typeof error != "undefined") {
+                ex = new qx.io.remote.exception.JsonRpc(
+                  error.code,
+                  error.message,
+                  {
+                    request : originalRequest.data,
+                    response: response
+                  }
+                );
+                // inform listeners
+                this.fireDataEvent("error", ex);
+                // reject the individual promise
+                return originalRequest.promise.reject(ex);
+              }
+              // we have our result!
+              else if (typeof result != "undefined") {
+                // inform listeners
+                this.fireDataEvent("success", result);
+                // resolve the individual promise
+                return originalRequest.promise.resolve(result);
+              }
+              // invalid response
+              ex = new qx.io.remote.exception.Transport(
+                qx.io.remote.RpcError.type.INVALID_DATA,
+                "Invalid jsonrpc data for "+ method + ": result is missing.",
+                {
+                  request : originalRequest.data,
+                  response: response
+                }
+              );
+              return originalRequest.promise.reject(ex);
+            }
+            // response is a server call or notification
+            if (typeof method != "undefined") {
+              this.fireDataEvent("request", response);
+            }
+          }, this);
+          // resolve the send() promise
+          return resolve(batch.length > 1 ? batch : batch[0]);
+        }, this);
+        // send the request!
+        req.send();
+      }, this);
+      return promise;
+    },
+
+    /**
+     * Sends the request immediately without queueing it.
+     * See {@link qx.io.remote.Rpc#addRequest} and {@link
+     * qx.io.remote.Rpc#send}.
+     *
+     * @param method  {String}
+     * @param params {Array}
+     * @return {qx.Promise}
+     */
+    sendRequest : function(method, params) {
+      var promise = this.addRequest(method, params);
+      this.send().catch(function (err) {
+        promise.reject(err);
+      });
+      return promise;
+    },
+
+    /**
+     * Sends a notification immediately without queueing it.
+     * See {@link qx.io.remote.Rpc#addNotification} and
+     * {@link qx.io.remote.Rpc#send}.
+     *
+     * @param method {String}
+     * @param params {Array|undefined}
+     */
+    sendNotification : function(method, params) {
+      this.addNotification(method, params);
```

Comment:
This is inconsistent. `sendRequest` sends it without queuing it but `sendNotification` queues it. I think it should not be queued here.

Reply:
It does the same thing - first queuing (1393/1409), then sending (1394/1410) the request/notification.

Reply:
Yes, sorry. I misread it. I wonder, though, if this is the best implementation? If you're in the process of building up a batch, would it be useful to be able to send a request or a notification that is unrelated to that batch, e.g., addRequest, addRequest, sendRequest, addRequest, send... which would end up actually sending the single request of `sendRequest` before the three-request batch with sent with `send`...

Reply:
Ok, you're sayint that a batch should be its own object, so building batches could be done separately? I like the idea conceptually, but can you think of a use case in which you would use it?

Reply:
Say I wanted to send updates only when I've accumulated 20 of them, but I also have some periodic request I need to send. At present, those updates, each a call to `addRequest`, would be sent along with my periodic request (a call to `sendRequest`) even though I really want only the periodic request to be sent at that moment, and desire the updates to be sent later once I've accumulated enough of them. (Granted, I could create my own array of them, but in that case, I would prefer to pass an array to a (non-existent) `sendRequests` method or something of the sort.)

Reply:
I am convinced, I will create a `qx.io.jsonrpc.Batch` object which can be used in such a case, plus the shorthand methods which send just one single call.

Reply:
Maybe like this:
`qx.io.jsonrpc.Request` -> serializes to JSON
`qx.io.jsonrpc.Notification` -> serializes to JSON
`qx.io.jsonrpc.Batch.add()` -> adds one of the above
`qx.io.jsonrpc.Http` -> instance, has a `send()` method which takes one of the three objects above as an argument, i.e. either a request, a notification or a batch. 
What do you think?

Reply:
Or should the client instance be generic and take a transport object as an injected dependency?

Reply:
> Maybe like this:
> `qx.io.jsonrpc.Request` -> serializes to JSON
> `qx.io.jsonrpc.Notification` -> serializes to JSON
> `qx.io.jsonrpc.Batch.add()` -> instance, adds one of the above
> `qx.io.jsonrpc.Http` -> client instance, has a `send()` method which takes one of the three objects above as an argument, i.e. either a request, a notification or a batch.
> What do you think?

I like that general idea, and I like the injected transport concept to go along with it.  You could then do something like:
```
let batch1 = new qx.io.jsonrpc.Batch();
let batch2 = new qx.io.jsonrpc.Batch();
batch1.add(new qx.io.jsonrpc.Request(...)); // can intermix requests and notifications
batch2.add(new qx.io.jsonrpc.Request(...)); // to multiple batches
batch1.add(new qx.io.jsonrpc.Notification(...));
batch2.add(new qx.io.jsonrpc.Notification(...));
let transport = new qx.io.jsonrpc.transport.Http(configMap);
transport.send(batch1); // send a specified batch
transport.send(new qx.io.jsonrpc.Request(...); // directly send a specific request or notification
transport.send(batch2);
```

Reply:
Cool. This and maybe some sugar object `const client = new qx.io.jsonrpc.Client("https://my.server.com/endpoint")` which auto-detects the transport type from the URL and has some convenience methods - but basically, you can work with the transports themselves.